cmd/queuebench/main.go

```diff
@@ -0,0 +1,223 @@
+		if err := bench.c.Run(ctx); err != nil {
+			log.Printf("consumer run ended with an error: %s", err)
+		}
+		consumptionduration = time.Since(consumptionstart)
+	}()
+
+	log.Printf("start producing, will produce for %s", cfg.duration)
+	productionstart := time.Now()
+	if err := produce(ctx, bench.p, bench.Topics[0], cfg.eventSize, cfg.duration); err != nil {
+		log.Printf("error while producing records: %s", err)
+	}
+	productionduration := time.Since(productionstart)
+
+	log.Println("production ended")
+
+	log.Println("waiting for consumer to fetch all records")
+	ticker := time.NewTicker(100 * time.Millisecond)
+	defer ticker.Stop()
+	timer := time.NewTimer(cfg.timeout)
+	defer timer.Stop()
+
+	var rm metricdata.ResourceMetrics
+	totalproduced := int64(0)
+	totalconsumed := int64(0)
+wait:
+	for {
+		select {
+		case <-ctx.Done():
+			log.Panic("context closed, terminating execution")
+		case <-timer.C:
+			log.Println("reached timeout, moving on")
+			break wait
+		case <-ticker.C:
+			if err := rdr.Collect(ctx, &rm); err != nil {
+				// NOTE: consider any error here as transient and don't trigger termination
+				log.Printf("cannot collect otel metrics: %s", err)
+				continue
+			}
+
+			totalproduced = getSumInt64Metric("github.com/elastic/apm-queue/kafka", "producer.messages.produced", rm)
+			totalconsumed = getSumInt64Metric("github.com/elastic/apm-queue/kafka", "consumer.messages.fetched", rm)
+			if totalconsumed >= totalproduced {
+				log.Println("consumption ended")
+				break wait
+			}
+		}
+	}
+	if err := bench.c.Close(); err != nil {
+		log.Panicf("error closing consumer: %s", err)
+	}
+
+	log.Println("==> benchmark ")
+	if err := bench.p.Close(); err != nil {
+		log.Panicf("error closing producer: %s", err)
+	}
+
+	duration := time.Since(start)
+	log.Printf("it took %s (-duration=%s)", duration, cfg.duration)
+	log.Printf("time spent producing: %s", productionduration)
+	log.Printf("time spent consuming: %s", consumptionduration)
+	log.Printf("total produced/consumed: %d/%d", totalproduced, totalconsumed)
+
+	log.Println("collecting metrics")
+	rdr.Collect(context.Background(), &rm)
+	if err = display(rm); err != nil {
+		log.Panicf("failed displaying metrics: %s", err)
+	}
+
+	if totalproduced != totalconsumed {
+		log.Panicf("total produced and consumed don't match: %d vs %d", totalproduced, totalconsumed)
+	}
+
+	log.Println("bench run completed successfully")
+}
+
+func produce(ctx context.Context, p *kafka.Producer, topic apmqueue.Topic, size int, duration time.Duration) error {
+	buf := make([]byte, size)
+
+	_, err := rand.Read(buf)
+	if err != nil {
+		return fmt.Errorf("cannot read random bytes: %w", err)
+	}
+
+	record := apmqueue.Record{
+		Topic: topic,
+		Value: buf,
+	}
+
+	deadline := time.Now().Add(duration)
+	for time.Now().Before(deadline) {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+		if err = p.Produce(ctx, record); err != nil {
+			return err
+		}
+	}
+
+	return err
```

Comment:
```suggestion
		if err := p.Produce(ctx, record); err != nil {
			return err
		}
	}

	return nil
```

The error will always be nil at the end of the loop. By keeping the `err` variables minimally scoped, we can prove that is the case, since then there will be no `err` in scope at the end of the loop.

Reply:
added in [edac17a](https://github.com/elastic/apm-queue/pull/256/commits/edac17a5bff07217794dfc87f91b9ad23c009f5e)